### PR TITLE
[JENKINS-44100] - Detach JNLP protocol management logic to a plugin

### PR DESCRIPTION
**Pull request by oleg-nenashev**
This is a "weekend hack project" attempt to move some boilerplate code outside the core in order to make it more maintainable and to allow applying changes to earlier core versions. Pull request on another side: https://github.com/oleg-nenashev/agent-remoting-support-plugin/pull/1

See [JENKINS-44100](https://issues.jenkins-ci.org/browse/JENKINS-44100).

Details:

- [x] [JENKINS-44100](https://issues.jenkins-ci.org/browse/JENKINS-44100) - Detach JNLP and CLI protocol handlers and aux stuff to the plugin
- [x] Provide new extension points, which allow retaining the original behavior of the core when the detached plugin is installed.
- [x] Add a last resort `NoOpComputerLauncher`, which is being used as a default fallback when the detached plugin is not installed
- [x] [JENKINS-44101](https://issues.jenkins-ci.org/browse/JENKINS-44101) - provide API for shipping custom Remoting.jar to agent endpoints like `jnlpJars/slave.jar` (to be implemented in the detached plugin)

TODOs:

- [ ] Decide which other components need to be detached. E.g. some bits of `TcpSlaveAgentListener`,  PingProtocol and Pinging logic, PingThread, etc. could be candidates
- [ ] Decide if the CLI stuff should go to another detached plugin (Remoting CLI Support?). It seems reasonable, because this mode is deprecated
- [ ] Hide `NoOpComputerLauncher` from UI
- [ ] Register the plugin as a detached plugin

### Changelog entries

Proposed changelog entries:

* Entry 1: TBD
* ...

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Link to JIRA ticket in description, if appropriate
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc)

### Desired reviewers

@reviewbybees, esp. @stephenc 

@mention

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Also CC @batmat, because some bits of https://github.com/jenkinsci/versioncolumn-plugin can probably be moved to this plugin

**Comment by jglick** [comment]
I do not see why the `CLIProtocol` and `CLIProtocol2` are included here. Seem unrelated to the rest of the change, which is clearly about JNLP agents. Yes they both happen to implement `AgentProtocol` and make use of `hudson.remoting.*` but that is about it.

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 1]
```diff
@@ -226,7 +227,14 @@ public String getIcon() {
...
     public boolean isJnlpAgent() {
-        return launcher instanceof JNLPLauncher;
+        final Class<?> jnlpLauncherClass;
+        try {
+            jnlpLauncherClass = Class.forName("hudson.slaves.JNLPLauncher");
+        } catch (Throwable ex) {
+            // Ignore
+            return false;
+        }
+        return jnlpLauncherClass.isAssignableFrom(launcher.getClass());
```
Comment (jglick) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: More simply:

```java
return getClass().getName().equals("hudson.slaves.JNLPLauncher");
```

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 2]
```diff
@@ -661,12 +669,20 @@ public void tryReconnect() {
...
+        return JnlpDataProvider.getJarFile(fileName);
     }
 
+    /**
+     * @deprecated 
+     */
     @WebMethod(name="slave-agent.jnlp")
     public HttpResponse doSlaveAgentJnlp(StaplerRequest req, StaplerResponse res) throws IOException, ServletException {
-        return new EncryptedSlaveAgentJnlpFile(this, "slave-agent.jnlp.jelly", getName(), CONNECT);
+        final HttpResponse jnlpFile = JnlpDataProvider.getJnlpFile(this, req);
```
Comment (jglick) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: I think we could just delete the web method and introduce an action with a matching URL in the paired PR.

File: core/src/main/java/jenkins/AgentProtocol.java [hunk 3]
```diff
@@ -23,6 +24,14 @@
  * @see TcpSlaveAgentListener
  */
 public abstract class AgentProtocol implements ExtensionPoint {
+    
+    /**
+     * This secret value is used as a seed for agents.
+     * @since TODO
+     */
+    public static final HMACConfidentialKey AGENT_SECRET =
```
Comment (jglick) [inline core/src/main/java/jenkins/AgentProtocol.java]: This should not exist.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 4]
```diff
@@ -3959,11 +3960,11 @@ public void doLogout( StaplerRequest req, StaplerResponse rsp ) throws IOExcepti
      * Serves jar files for JNLP agents.
      */
     public Slave.JnlpJar getJnlpJars(String fileName) {
-        return new Slave.JnlpJar(fileName);
+        return JnlpDataProvider.getJarFile(fileName);
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: Why are we moving this? Seems unrelated to the rest of the PR.

File: core/src/main/java/jenkins/slaves/DefaultComputerLauncherProvider.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.model.Node;
+import hudson.slaves.ComputerLauncher;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+
+/**
+ * Allows defining default computer launchers.
```
Comment (jglick) [inline core/src/main/java/jenkins/slaves/DefaultComputerLauncherProvider.java]: I do not think this needs to exist. We just look for the first entry in `ExtensionList<ComputerLauncher>`.

**Comment by oleg-nenashev** [comment]
@jglick 

> I do not see why the CLIProtocol and CLIProtocol2 are included here. Seem unrelated to the rest of the change, which is clearly about JNLP agents. Yes they both happen to implement AgentProtocol and make use of hudson.remoting.* but that is about it.

I agree, see my TODO list. I consider creating a detached "Remoting CLI Plugin", which would rely on the "Agent Remoting Support Plugin". This dependency is not ideal, but IMHO it is fine since this CLI Plugin immediately becomes "deprecated"

File: core/src/main/java/jenkins/slaves/JnlpDataProvider.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+import hudson.model.Slave;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.StaplerRequest;
+
+/**
+ * Allows providing custom data for JNLP agents.
+ * This extension point is mainly created for the Agent Remoting Support plugin.
+ * It can be also used in other plugins in order to get custom behavior.
```
Comment (jglick) [inline core/src/main/java/jenkins/slaves/JnlpDataProvider.java]: I do not think this API needs to exist.

File: core/src/main/java/jenkins/slaves/NoOpComputerLauncher.java [hunk 7]
```diff
@@ -23,49 +23,28 @@
...
-
-    public IOHubProvider() {
-        try {
-            hub = IOHub.create(Computer.threadPoolForRemoting);
-        } catch (IOException e) {
-            LOGGER.log(Level.SEVERE, "Failed to launch IOHub", e);
-            this.hub = null;
-        }
-    }
+@Restricted(NoExternalUse.class)
```
Comment (jglick) [inline core/src/main/java/jenkins/slaves/NoOpComputerLauncher.java]: Hmm, overzealous rename detection from Git?

File: core/src/main/java/jenkins/slaves/NoOpComputerLauncher.java [hunk 8]
```diff
@@ -23,49 +23,28 @@
...
 
-    @Terminator
-    public void cleanUp() throws IOException {
-        if (hub != null) {
-            hub.close();
-            hub = null;
-        }
+    @Override
+    public void launch(SlaveComputer computer, TaskListener listener) throws IOException, InterruptedException {
+        throw new IOException("Cannot launch computer " + computer + ", because a default NoOpComputerLauncher is specified");
```
Comment (jglick) [inline core/src/main/java/jenkins/slaves/NoOpComputerLauncher.java]: Use `AbortException`.

File: core/src/main/java/jenkins/slaves/NoOpComputerLauncher.java [hunk 9]
```diff
@@ -23,49 +23,28 @@
...
-import java.util.logging.Logger;
-import org.jenkinsci.remoting.protocol.IOHub;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
 
 /**
- * Singleton holder of {@link IOHub}
- *
- * @since 2.27
+ * Computer Launcher, which does nothing.
```
Comment (jglick) [inline core/src/main/java/jenkins/slaves/NoOpComputerLauncher.java]: I do not think this really needs to exist. Allow `SlaveComputer.launcher` to be null and handle that directly when trying to call the launcher.

File: test/src/test/java/hudson/tools/InstallerTranslatorTest.java [hunk 10]
```diff
@@ -52,7 +52,8 @@
 
     @Issue("JENKINS-23517")
     @Test public void offlineNodeForJDK() throws Exception {
-        Node slave = new DumbSlave("disconnected-slave", null, "/wherever", "1", Node.Mode.NORMAL, null, new JNLPLauncher(), RetentionStrategy.NOOP, Collections.<NodeProperty<?>>emptyList());
+        Node slave = new DumbSlave("disconnected-slave", null, "/wherever", "1", Node.Mode.NORMAL, null, 
+                DefaultComputerLauncherProvider.findDefaultLauncher(DumbSlave.class), RetentionStrategy.NOOP, Collections.<NodeProperty<?>>emptyList());
```
Comment (jglick) [inline test/src/test/java/hudson/tools/InstallerTranslatorTest.java]: This makes no sense. Leave the test alone and add the new plugin as a `test`-scoped dependency. Or, switch to `CommandLauncher`.

File: test/src/test/java/jenkins/security/Security218Test.java [hunk 11]
```diff
@@ -80,7 +80,8 @@ public Object call() {
      * @see #launchJnlpSlave(Slave)
      */
     public DumbSlave createJnlpSlave(String name) throws Exception {
-        DumbSlave s = new DumbSlave(name, "", System.getProperty("java.io.tmpdir") + '/' + name, "2", Mode.NORMAL, "", new JNLPLauncher(), RetentionStrategy.INSTANCE, Collections.EMPTY_LIST);
+        DumbSlave s = new DumbSlave(name, "", System.getProperty("java.io.tmpdir") + '/' + name, "2", Mode.NORMAL, "", 
+                DefaultComputerLauncherProvider.findDefaultLauncher(DumbSlave.class), RetentionStrategy.INSTANCE, Collections.EMPTY_LIST);
```
Comment (jglick) [inline test/src/test/java/jenkins/security/Security218Test.java]: Ditto.

**Review by jglick — DISMISSED** [review]
Too much API debris being left behind in core. If you are going to move the JNLP launcher to a plugin, make a clean break of it.

**Comment by jglick** [comment]
> creating a detached "Remoting CLI Plugin"

Why bother? The code is deprecated, it can just sit in core unless and until we find a way to separate the whole CLI system to a plugin (already a filed task but tricky).

File: core/src/main/java/hudson/slaves/SlaveComputer.java (hunk 1 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: No since JNLPLauncher is not final

File: core/src/main/java/hudson/model/Slave.java [hunk 12]
```diff
@@ -348,9 +349,14 @@ public FilePath getRootPath() {
     /**
      * Web-bound object used to serve jar files for JNLP.
      */
-    public static final class JnlpJar implements HttpResponse {
+    public static class JnlpJar implements HttpResponse {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Slave.java]: It is for https://issues.jenkins-ci.org/browse/JENKINS-44101, see the TODO in the paired PR

File: core/src/main/java/hudson/slaves/SlaveComputer.java (hunk 2 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: Binary compatibility :(

File: core/src/main/java/jenkins/AgentProtocol.java (hunk 3 above)
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/AgentProtocol.java]: I am open to proposals

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 4 above)
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: https://issues.jenkins-ci.org/browse/JENKINS-44101

File: core/src/main/java/jenkins/slaves/DefaultComputerLauncherProvider.java (hunk 5 above)
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/slaves/DefaultComputerLauncherProvider.java]: There is no guarantee there will be a default constructor for it. And it will break the original behavior since JNLPLauncher() is expected to be a default behavior i n the code, including readResolve()

File: core/src/main/java/jenkins/slaves/NoOpComputerLauncher.java (hunk 7 above)
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/slaves/NoOpComputerLauncher.java]: yep. New class

**Comment by oleg-nenashev** [comment]
> Why bother? The code is deprecated, it can just sit in core unless and until we find a way to separate the whole CLI system to a plugin (already a filed task but tricky).

The code depends on `NioChannelSelector`, which I would like to move to the detached plugin. Will have to leave it in the core otherwise

File: core/src/main/java/hudson/slaves/SlaveComputer.java (hunk 1 above)
Comment (jglick) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: I know, but there are no expected subclasses. Anyway this is good enough if verbose.

File: core/src/main/java/hudson/slaves/SlaveComputer.java (hunk 2 above)
Comment (jglick) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: For a web method? No need. There should be no callers. If there are some in private plugins, those should get fixed.

File: core/src/main/java/jenkins/AgentProtocol.java (hunk 3 above)
Comment (jglick) [inline core/src/main/java/jenkins/AgentProtocol.java]: Just delete. All uses should be present in the plugin only.

File: core/src/main/java/hudson/slaves/SlaveComputer.java [hunk 13]
```diff
@@ -172,7 +173,7 @@ public boolean isAcceptingTasks() {
      * @since 1.498
      */
     public String getJnlpMac() {
-        return JnlpSlaveAgentProtocol.SLAVE_SECRET.mac(getName());
+        return AgentProtocol.AGENT_SECRET.mac(getName());
```
Comment (jglick) [inline core/src/main/java/hudson/slaves/SlaveComputer.java]: (Or simply deleted.)

File: core/src/main/java/hudson/model/Slave.java (hunk 12 above)
Comment (jglick) [inline core/src/main/java/hudson/model/Slave.java]: I am not sure that use case is a good idea, but even if it is, you do not need to do this. The JNLP file produced by the plugin can simply point to a custom endpoint which serves a different JAR.

File: core/src/main/java/jenkins/slaves/DefaultComputerLauncherProvider.java (hunk 5 above)
Comment (jglick) [inline core/src/main/java/jenkins/slaves/DefaultComputerLauncherProvider.java]: > There is no guarantee there will be a default constructor for it.

Define a `newInstance` method in the descriptor type.

> it will break the original behavior since JNLPLauncher() is expected to be a default behavior i n the code

So? That behavior can change. The `readResolve` is unimportant and there should be no other calls I think.

**Comment by jglick** [comment]
> The code depends on `NioChannelSelector`, which I would like to move

Ah, got it. OK, so move it and leave it deprecated, no need to ever move it elsewhere.

File: core/src/main/java/hudson/model/Slave.java (hunk 12 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Slave.java]: Yeah. But the idea is to tweak the default endpoint. In such case there will be no changes required on the agent side, docker images, documentation, etc.

File: core/src/main/java/jenkins/slaves/NoOpComputerLauncher.java (hunk 9 above)
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/slaves/NoOpComputerLauncher.java]: Maybe

File: test/src/test/java/hudson/tools/InstallerTranslatorTest.java (hunk 10 above)
Comment (oleg-nenashev) [inline test/src/test/java/hudson/tools/InstallerTranslatorTest.java]: I just wanted to get some free coverage of the extension point. But yeah, maybe it is a bad idea

**Comment by oleg-nenashev** [comment]
This pull request is not recoverable, somebody needs to do it from scratch after lots of related code changes. I do not anticipate to get time for such Remoting-related work in next 2-3 months, so I am just closing it.

https://issues.jenkins-ci.org/browse/JENKINS-44100 has been updated to reflect the state and to point to pull requests

**Comment by jglick** [comment]
> creating a detached "Remoting CLI Plugin"

Simpler to delete without replacement! https://github.com/jenkinsci/jenkins/pull/3838
